src/Analyzer.JsonRuleEngine/Evaluation.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Azure.Templates.Analyzer.RuleEngines.JsonEngine
+{
+    /// <summary>
+    /// Describes the result of an evaluation of an expression.
+    /// </summary>
+    internal class Evaluation
+    {
+        /// <summary>
+        /// Gets a value indicating whether or not the rule for this evaluation passed.
+        /// </summary>
+        public bool Passed { get; internal set; }
+
+        /// <summary>
+        /// A collections of results from this evaluation
+        /// </summary>
+        public IEnumerable<JsonRuleResult> Results { get; set; }
+
+        private IEnumerable<JsonRuleResult> resultsEvaluatedTrue;
+        private IEnumerable<JsonRuleResult> resultsEvaluatedFalse;
+
+        /// <summary>
+        /// Creates an <see cref="Evaluation"/>.
+        /// </summary>
+        /// <param name="passed">Determines whether or not the rule for this evaluation passed.</param>
+        /// <param name="results"><see cref="IEnumerable"/> of results.</param>
+        public Evaluation(bool passed, IEnumerable<JsonRuleResult> results) => (this.Passed, this.Results) = (passed, results);
+
+        /// <summary>
+        /// Gets all the results that evaluated to true.
+        /// </summary>
+        /// <returns>The results that evaluated to true.</returns>
+        public IEnumerable<JsonRuleResult> GetResultsEvaluatedTrue()
+        {
+            if (resultsEvaluatedTrue == null)
+            {
+                resultsEvaluatedTrue = Results.ToList().FindAll(r => r.Passed);
```

Comment:
There's a hidden bug here. :)
When this class is created, if `results` comes from the results of a function that uses `yield return` (most evaluations do), whichever of `GetResultsEvaluatedTrue` or `GetResultsEvaluatedFalse` is called first will iterate through the results and throw away some of them, and then the other function will not find anything.

You can test this by creating the results in the `Evaluate_TwoLeafExpressions_ExpectedResultIsReturned` test from `yield return` statements instead of any array.

I can explain in more detail offline if you want.